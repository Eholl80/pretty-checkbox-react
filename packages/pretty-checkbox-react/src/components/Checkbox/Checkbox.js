// @flow

// https://www.w3.org/TR/wai-aria-practices-1.1/#checkbox

import * as React from 'react';
import classNames from 'classnames';
import { warn, mergeRefs, isValidIconType, error, removeIndexFromArray, isBoolean, isChecked } from '../../utils/utils';

import { Icon } from '../Icon/Icon';
import { BaseInput } from '../BaseInput/BaseInput';

import {
    type PCRColors,
    type PCRColorsOutline,
    type RadioCheckboxVariants,
    type RadioCheckboxFillModes,
    type RadioCheckboxAnimations,
    type IconPropType,
} from '../../types';

export type CheckboxState = boolean | 'indeterminate' | any[];

export type CheckboxProps = {
    /**
     * Specify one of the colors.
     */
    color?: PCRColors | PCRColorsOutline,

    /**
     * Specify one of the variants.
     */
    variant?: RadioCheckboxVariants,

    /**
     * Specify one of the fill modes.
     */
    fill?: RadioCheckboxFillModes,

    /**
     * Specify one of the animations.
     */
    animation?: RadioCheckboxAnimations,

    /**
     * Controlled mode checkbox.
     */
    checked?: boolean,

    /**
     * Pass an icon to render.
     */
    icon?: IconPropType,

    /**
     * Runs inside the default onChange handler for the Checkbox.
     *
     * Optionally return `false` from a synchronous action to prevent the
     * checkbox from being un/checked.
     */
    onChange?: (e: SyntheticMouseEvent<HTMLInputElement>) => ?boolean,

    /**
     * Set the checkbox as disabled.
     */
    disabled?: boolean,

    /**
     * Set the checkbox as locked. This will also set the
     * `aria-disabled` attribute to `true`.
     */
    locked?: boolean,

    /**
     * Pass additional class names to the `.pretty` div.
     */
    className?: string,

    /**
     * The `state` prop from` useCheckboxState`.
     */
    state?: CheckboxState,

    /**
     * The `setState` dispatcher from `useCheckboxState`.
     */
    setState?: (state: CheckboxState) => void,

    /**
     * Stub...
     */
    value?: string,

    /**
     * Additional HTML attributes to pass to the `input` element.
     * TODO: fix flow type after https://github.com/facebook/flow/pull/7569 ðŸ™„
     */
    htmlProps?: Object,

    /**
     * Text to show inside the label _or_ a render function
     * that allows you to render the internal pretty-checkbox
     * `.state` div.
     *
     * @example
     * <Checkbox>
     *     {({ color, icon }) => (
     *         <div className="state my-custom-class-name">
     *             {icon}
     *             <label>Something super special</label>
     *         </div>
     *     )}
     * </Checkbox>
     */
    children: React.Node | ((...props: Object) => React.Node),
};

export const useCheckboxState = (initialState: CheckboxState | (() => CheckboxState) = false) => {
    const [state, setState] = React.useState<CheckboxState>(initialState);

    return { state, setState };
};

/**
 * You probably shouldn't be using this ;)
 */
export const __handleCheckboxChange = ({
    setState,
    state,
    value,
    bool,
}: {
    setState: (state: CheckboxState) => void,
    state?: CheckboxState | string,
    value?: string,
    bool: boolean,
}) => {
    if (bool) {
        setState(!state);
    } else {
        const array = Array.isArray(state) ? state : [];
        const index = array.indexOf(value);

        if (index === -1) {
            setState([...array, value]);
        } else {
            setState([...removeIndexFromArray(array, index)]);
        }
    }
};

export const Checkbox = React.forwardRef<CheckboxProps, HTMLInputElement>(
    (
        {
            color,
            variant,
            fill,
            animation,
            checked: checkedProp,
            onChange,
            disabled,
            locked,
            children,
            className,
            icon = { type: null, icon: null },
            value,
            state,
            setState,
            ...htmlProps
        },
        ref
    ) => {
        const bool = isBoolean(value);
        const checked = isChecked(checkedProp, state, value);

        const checkboxRef = React.useRef(mergeRefs(ref, React.createRef()));

        const handleChange = React.useCallback(
            (e: SyntheticMouseEvent<HTMLInputElement>) => {
                if (disabled || locked) {
                    return;
                }

                if (typeof onChange === 'function') {
                    if (onChange(e) === false) {
                        return;
                    }
                }

                if (typeof setState !== 'function') {
                    return;
                }

                __handleCheckboxChange({ setState, state, value, bool });
            },
            [setState, state, disabled, locked, onChange]
        );

        React.useEffect(() => {
            if (!checkboxRef.current && state === 'indeterminate') {
                warn(
                    'Checkbox',
                    "Can't set indeterminate state because `ref` wasn't passed to component."
                );

                return;
            }

            if (checkboxRef.current) {
                if (state === 'indeterminate') {
                    // $FlowFixMe
                    checkboxRef.current.indeterminate = true;
                    // $FlowFixMe
                } else if (checkboxRef.current.indeterminate) {
                    // $FlowFixMe
                    checkboxRef.current.indeterminate = false;
                }
            }
        }, [state]);

        return (
            <BaseInput
                // expected props
                type="checkbox"
                role="checkbox"
                icon={icon}
                color={color}
                locked={locked}
                ref={checkboxRef.current}
                // htmlProps
                disabled={disabled}
                aria-checked={state === 'indeterminate' ? 'mixed' : checked}
                aria-disabled={disabled || locked}
                // $FlowFixMe
                onChange={handleChange}
                checked={checked}
                value={value}
                tabIndex="0"
                className={classNames(
                    {
                        'p-default': !icon.type && (!animation || animation === 'smooth'),
                        [variant ? `p-${variant}`: '']: variant,
                        [animation ? `p-${animation}` : '']: animation,
                        [icon.type ? `p-${icon.type}` : '']: icon.type,
                        [fill ? `p-${fill}`: '']: fill,
                    },
                    className
                )}>
                {children}
            </BaseInput>
        );
    }
);

Checkbox.displayName = 'Checkbox';
