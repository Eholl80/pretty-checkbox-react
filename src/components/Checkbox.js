// @flow

import React from 'react';
import classNames from 'classnames';
import { getBaseClassName } from '../utils/utils';

import Input, { PREFIX } from './Input';
import type { InputProps } from './Input';

export type CheckboxProps = {
    ...InputProps,

    /**
     * Set true when the checkbox is in an indeterminate state.
     */
    indeterminate?: boolean,
};

function Checkbox(props: CheckboxProps) {
    const { animation, className, ...rest } = props;

    if (
        animation &&
        animation !== 'smooth' &&
        animation !== 'pulse' &&
        !props.icon &&
        !props.image &&
        !props.svg
    ) {
        throw new Error(
            `animation '${animation}' is incompatible with default checkbox styles. You must specify an icon, image, or a svg.`
        );
    }

    return (
        <Input
            type="checkbox"
            className={classNames(
                // $ExpectError
                getBaseClassName(props, PREFIX),
                props.indeterminate ? 'p-has-indeterminate' : null,
                className
            )}
            animation={animation}
            {...rest}
        />
    );
}

export default Checkbox;
