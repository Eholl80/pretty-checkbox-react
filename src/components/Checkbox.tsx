import * as React from 'react';

import { useIcon } from '../hooks/useIcon';
import { CommonCheckboxRadioProps } from '../types/CommonProps';
import { PrettyComponent } from '../factory/PrettyComponent';

export type CheckboxState = boolean | any[] | 'indeterminate';

export type CheckboxProps = Omit<CommonCheckboxRadioProps<CheckboxState>, 'type'> & {
    type?: 'checkbox';
};

export const useCheckboxState = ({
    initialState = false,
}: { initialState?: CheckboxState } = {}) => {
    const [state, setState] = React.useState<CheckboxState>(initialState);

    return {
        state,
        setState,
        onChange: React.useCallback((e: React.ChangeEvent<HTMLInputElement>) => {
            const value = e.currentTarget.value;

            setState(state => {
                if (Array.isArray(state)) {
                    const index = state.indexOf(value);

                    if (index === -1) {
                        state.push(value);
                    } else {
                        state.splice(index, 1);
                    }

                    return state.slice();
                }

                return !state;
            });
        }, []),
    };
};

export const Checkbox: React.FC<CheckboxProps> = ({
    value: userValue,
    icon: userIcon,
    ...rest
}: CheckboxProps) => {
    const { icon, iconType } = useIcon(userIcon);
    const value = typeof userValue === 'undefined' ? '' : userValue;

    return React.createElement(PrettyComponent, {
        type: 'checkbox',
        icon,
        iconType,
        value,
        ...rest,
    });
};

Checkbox.displayName = 'Checkbox';
