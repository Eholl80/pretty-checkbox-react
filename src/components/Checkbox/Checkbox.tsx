import * as React from 'react';

import { useCheckboxChange } from './hooks/useCheckboxChange';
import { mergeRefs, warn, getChecked } from '../../utils/utils';
import { BaseProps } from '../../types/BaseProps';
import { RadioCheckboxFillModes } from '../../types/types';
import { BaseInput } from '../BaseInput/BaseInput';
import { useGenericState } from '../../hooks/useGenericState';

export type CheckboxState = boolean | 'indeterminate' | any[] | undefined;

export type CheckboxProps = BaseProps<CheckboxState> & {
    /**
     * Specify one of the fill modes.
     */
    fill?: RadioCheckboxFillModes;
};

export const useCheckboxState = (initialState: CheckboxState = false) =>
    useGenericState<CheckboxState>(initialState);

export const Checkbox = React.forwardRef(
    (
        { state, setState, value, disabled, locked, onChange, ...rest }: CheckboxProps,
        htmlRef: React.Ref<HTMLInputElement>
    ) => {
        const ref = React.useRef<HTMLInputElement>(null);

        const defaultOnChange = useCheckboxChange({
            setState,
            value,
            disabled,
            locked,
        });

        React.useLayoutEffect(() => {
            if (!ref.current && state === 'indeterminate') {
                warn(
                    'Checkbox',
                    "Can't set indeterminate state because `ref` wasn't passed to component."
                );

                return;
            }

            if (ref.current) {
                if (state === 'indeterminate') {
                    ref.current.indeterminate = true;
                } else if (ref.current.indeterminate) {
                    ref.current.indeterminate = false;
                }
            }
        }, [state]);

        // TODO: need to add htmlFor generated ID by default
        // allow consumer to provide their own as an escape

        return (
            <BaseInput
                state={getChecked<CheckboxState>({ value, state, setState })}
                setState={setState}
                ref={mergeRefs(htmlRef, ref)}
                disabled={disabled}
                locked={locked}
                value={value}
                onChange={onChange || defaultOnChange}
                {...rest}
            />
        );
    }
);

Checkbox.displayName = 'Checkbox';
