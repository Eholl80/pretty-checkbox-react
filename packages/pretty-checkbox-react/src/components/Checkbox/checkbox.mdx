---
name: Checkbox
---

import { Checkbox, useCheckboxState } from './Checkbox';
import { Playground, Props } from 'docz';

# Checkbox

<Props of={Checkbox} />

Pretty Checkbox React (PCR) is flexible and allows you to completely control the state of the checkbox yourself.
While there are a few ways of doing this, we recommend using the `useCheckboxState` hook to make this happen.
Of course, you may control checkbox state using `checked` and `onChange` props if you fancy too.

## Basic Usage with `useCheckboxState`

<Playground>
    {() => {
        const checkbox = useCheckboxState();

        return (
            <Checkbox {...checkbox}>
                Basic Checkbox with Hooks
            </Checkbox>
        );
    }}
</Playground>

## Basic Usage with Hooks

<Playground>
    {() => {
        const [checked, dispatch] = React.useReducer(state => !state, false);

        return (
            <Checkbox fill="thick" checked={checked} onChange={dispatch}>
                Controlled
            </Checkbox>
        );
    }}

</Playground>

## Basic Usage without Hooks

Not on the hooks train yet? No worries üòâ

<Playground>
    {class MyCheckbox extends React.Component {
            constructor() {
                this.state = { checked: false };
                this.toggleCheckbox = this.toggleCheckbox.bind(this);
            }

            toggleCheckbox() {
                this.setState({ checked: !this.state.checked });
            }

            render() {
                return (
                    <Checkbox
                        variant="curve"
                        checked={this.state.checked}
                        onChange={this.toggleCheckbox}>
                        Basic Checkbox without Hooks
                    </Checkbox>
                );
            }
        }}

</Playground>

## Indeterminate

Indeterminate, or tri-state, checkboxes in React are a source of pain. No, really. It's a [legitimate problem](https://github.com/facebook/react/issues/1798).
Thankfully PCR takes care of that for you. PCR internally creates a `ref` on you behalf. You can always supply your own `ref`
if you need extra control and we'll use that instead!

<Playground>
    {() => {
        const checkbox = useCheckboxState();

        const handleMakeIndeterminate = React.useCallback(() => {
            checkbox.setState('indeterminate');
        }, [checkbox.setState]);

        return (
            <>
                <Checkbox
                    variant="curve"
                    icon={{
                        icon: (
                            <i className="material-icons">
                                {checkbox.state !== 'indeterminate'
                                    ? 'check'
                                    : 'remove'}
                            </i>
                        ),
                        type: 'icon'
                    }}
                    {...checkbox}>
                    Indeterminate
                </Checkbox>
                <button onClick={handleMakeIndeterminate}>Apply</button>
            </>
        );
    }}
</Playground>

Toy examples aside, here's something a tad more realistic:

<Playground>
    {() => {
        // example lifted from https://reakit.io/docs/checkbox/
        function useTreeState({ values }) {
            const group = useCheckboxState();
            const items = useCheckboxState();

            React.useEffect(() => {
                if (group.state === true) {
                    items.setState(values);
                } else if (group.state === false) {
                    items.setState([]);
                }
            }, [group.state]);

            React.useEffect(() => {
                if (items.state.length === values.length) {
                    group.setState(true);
                } else if (items.state.length) {
                    group.setState('indeterminate');
                } else {
                    group.setState(false);
                }
            }, [items.state]);

            return { group, items };
        }

        const Example = () => {
            const values = ['Apple', 'Orange', 'Watermelon'];
            const { group, items } = useTreeState({ values });

            return (
                <ul>
                    <li>
                        <Checkbox
                            {...group}
                            icon={{
                                icon: (
                                    <i className="material-icons">
                                        {group.state !== 'indeterminate'
                                            ? 'check'
                                            : 'remove'}
                                    </i>
                                ),
                                type: 'icon',
                            }}>
                            Fruits
                        </Checkbox>
                    </li>
                    <ul>
                        {values.map((value, i) => (
                            <li key={i}>
                                <Checkbox
                                    {...items}
                                    value={value}
                                    icon={{ icon: <i className="material-icons">check</i>, type: 'icon' }}>
                                    {value}
                                </Checkbox>
                            </li>
                        ))}
                    </ul>
                </ul>
            );
        };

        return <Example />;
    }}
</Playground>

## Checkbox Variants

The awesome underlying CSS library that powers PCR offers various combinations of fill modes and variants. Here's a comprenhesive list:

### Square Checkbox

<Playground>
    <>
        <Checkbox checked>Default</Checkbox>
        <Checkbox fill="fill" checked>
            Fill
        </Checkbox>
        <Checkbox fill="thick" checked>
            Thick
        </Checkbox>
    </>
</Playground>

### Curved Checkbox

<Playground>
    <>
        <Checkbox variant="curve" checked>
            Default
        </Checkbox>
        <Checkbox variant="curve" fill="fill" checked>
            Fill
        </Checkbox>
        <Checkbox variant="curve" fill="thick" checked>
            Thick
        </Checkbox>
    </>
</Playground>

### Round Checkbox

<Playground>
    <>
        <Checkbox variant="round" checked>
            Default
        </Checkbox>
        <Checkbox variant="round" fill="fill" checked>
            Fill
        </Checkbox>
        <Checkbox variant="round" fill="thick" checked>
            Thick
        </Checkbox>
    </>
</Playground>

## Disabled and Locked Checkbox

<Playground>
    <>
        <Checkbox variant="round" checked locked>
            Locked
        </Checkbox>
        <Checkbox variant="round" fill="fill" checked disabled>
            Disabled
        </Checkbox>
    </>
</Playground>

## Animations

The CSS library that powers PCR comes with lovely in-house animatons, five to be exact.

* smooth
* pulse
* jelly - requires an icon
* tada - requires an icon
* rotate - requires an icon

<Playground>
    {() => {
        const items = useCheckboxState([]);

        return (
            <>
                <Checkbox animation="smooth" value="smooth" {...items}>
                    Smooth
                </Checkbox>
                <Checkbox
                    animation="jelly"
                    variant="round"
                    value="jelly"
                    icon={{
                        type: 'icon',
                        icon: <i className="material-icons">check</i>
                    }}
                    color="primary"
                    {...items}>
                    Jelly
                </Checkbox>
                <Checkbox
                    animation="tada"
                    value="tada"
                    color="danger"
                    icon={{
                        type: 'icon',
                        icon: <i className="material-icons">whatshot</i>
                    }}
                    {...items}>
                    Tada
                </Checkbox>
                <Checkbox
                    animation="pulse"
                    value="pulse"
                    color="warning"
                    icon={{
                        type: 'icon',
                        icon: <i className="material-icons">wb_sunny</i>
                    }}
                    {...items}>
                    Pulse
                </Checkbox>
                <Checkbox
                    animation="rotate"
                    value="rotate"
                    icon={{
                        type: 'icon',
                        icon: <i className="material-icons">rotate_right</i>
                    }}
                    {...items}>
                    Rotate
                </Checkbox>
            </>
        );
    }}
</Playground>


## Custom Rendering

Need something custom? No worries üëç. We got your back. You can supply a child render function to any component and have full
control over the `.state` element of [pretty checkbox](https://lokesh-coder.github.io/pretty-checkbox/).

As a convenience to you, PCR will give you the following props:

-   `color: string`
-   `icon: { type: 'image' | 'svg' | 'icon' | null, icon: React.Node | null }`

```jsx
// node is the svg, icon, or image node
<Checkbox>
    {({ color, icon }) => (
        <div className={`state p-${color}`}>
            <label>My custom label</label>
            {icon.icon}
        </div>
    )}
</Checkbox>
```
