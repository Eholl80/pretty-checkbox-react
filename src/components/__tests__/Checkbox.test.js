import React from 'react';

import { render, cleanup, getByTestId, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import Checkbox from '../Checkbox';

afterEach(cleanup);

describe('Checkbox tests', function () {
    describe('Basic Checkbox usage', function () {
        it('should behave as a checkbox', function () {
            const handleChange = jest.fn();
            const { container } = render(<Checkbox className="dummy-selector" id="foo" onChange={handleChange} />);

            expect(getByTestId(container, 'pcr-input')).toHaveAttribute('type', 'checkbox');
            expect(getByTestId(container, 'pcr-wrapper')).toHaveClass('dummy-selector');

            fireEvent.click(getByTestId(container, 'pcr-input'));

            expect(handleChange).toHaveBeenCalled();
            expect(handleChange).toHaveBeenCalledTimes(1);
            expect(getByTestId(container, 'pcr-input').checked).toBe(true);

            fireEvent.click(getByTestId(container, 'pcr-input'));

            expect(handleChange).toHaveBeenCalledTimes(2);
            expect(getByTestId(container, 'pcr-input').checked).toBe(false);
            expect(container).toMatchSnapshot();
        });

        it('should render icons with the correct classNames', function () {
            let c1 = render(<Checkbox icon={<i className="mdi mdi-check" />}>Hello there.</Checkbox>).container;
            let c2 = render(<Checkbox svg={<i className="mdi mdi-check" />}>Hello there.</Checkbox>).container;
            let c3 = render(<Checkbox image={<i className="mdi mdi-check" />}>Hello there.</Checkbox>).container;

            expect(getByTestId(c1, 'pcr-wrapper')).toHaveClass('p-icon');
            expect(getByTestId(c2, 'pcr-wrapper')).toHaveClass('p-svg');
            expect(getByTestId(c3, 'pcr-wrapper')).toHaveClass('p-image');
        });
    });

    describe('Checkbox errors', function () {
        it('should throw when using an unsupported animation', function () {
            expect(() => render(<Checkbox animation="tada">Hello there.</Checkbox>)).toThrow();
        });
    });

    describe('Checkbox supports animations', function () {
        it('should have the correct smooth className', function () {
            const { container } = render(<Checkbox shape="round" color="primary" animation="smooth">Monday</Checkbox>);
            expect(getByTestId(container, 'pcr-wrapper')).toHaveClass('p-smooth');
        });

        it('should have the correct jelly className', function () {
            const { container } = render(<Checkbox shape="round" color="primary" icon={<i className="mdi mdi-check" />} animation="jelly">Interested</Checkbox>);
            expect(getByTestId(container, 'pcr-wrapper')).toHaveClass('p-jelly');
        });

        it('should have the correct tada className', function () {
            const { container } = render(<Checkbox shape="round" color="primary-o" icon={<i className="mdi mdi-heart" />} animation="tada">Good</Checkbox>);
            expect(getByTestId(container, 'pcr-wrapper')).toHaveClass('p-tada');
        });

        it('should have the correct rotate className', function () {
            const { container } = render(<Checkbox color="success" icon={<i className="mdi mdi-check" />} animation="rotate">Friends</Checkbox>);
            expect(getByTestId(container, 'pcr-wrapper')).toHaveClass('p-rotate');
        });

        it('should have the correct pulse className', function () {
            const { container } = render(<Checkbox color="warning-o" animation="pulse">Family</Checkbox>);
            expect(getByTestId(container, 'pcr-wrapper')).toHaveClass('p-pulse');
        });
    });

    describe('Checkbox indeterminate', function () {
        it('should provide style hooks when checkbox is an indeterminate', function () {
            const { container } = render(<Checkbox indeterminate>Hello there.</Checkbox>);

            expect(getByTestId(container, 'pcr-wrapper')).toHaveClass('p-has-indeterminate');
        });
    });

    describe('Checkbox ref tests', function() {
        it('should accept custom refs', function () {
            const ref = React.createRef();

            render(<Checkbox ref={ref} />);
            expect(ref.current).toBeDefined();
            expect(ref.current).toBeInstanceOf(HTMLInputElement);
        });
    });
});
