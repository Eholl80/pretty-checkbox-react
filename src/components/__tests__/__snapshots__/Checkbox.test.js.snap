// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox tests Basic Checkbox usage should behave as a checkbox 1`] = `
<div>
  <div
    class="pretty p-default"
    data-testid="pcr-wrapper"
  >
    <input
      data-testid="pcr-input"
      id="foo"
      type="checkbox"
      value=""
    />
    <div
      class="state"
      data-testid="pcr-state"
    />
  </div>
</div>
`;
